torch/_inductor/codegen/cpp.py

```diff
@@ -3619,7 +3738,198 @@ def codegen_nodes(self, nodes: List[SchedulerNode]):
         cpp_kernel_proxy = CppKernelProxy(kernel_group)
         cpp_kernel_proxy.codegen_nodes(nodes)
 
-        kernel_group.finalize_kernel(cpp_kernel_proxy, nodes)
+        def _can_fuse_outer_loop(
+            _outer_loop_fusion: OuterLoopFusions,
+            _cpp_kernel_proxy: CppKernelProxy,
+            _nodes: List[SchedulerNode],
+        ):
+            def _check_node_dependency(
+                __outer_loop_fusion: OuterLoopFusions,
+                __nodes: List[SchedulerNode],
+            ):
+                # List[SchedulerNode]
+                flatten_lazy_nodes_list, _ = torch.utils._pytree.tree_flatten(
+                    __outer_loop_fusion._lazy_nodes_list,
+                )
+                assert all(
+                    isinstance(_lazy_node, SchedulerNode)
+                    for _lazy_node in flatten_lazy_nodes_list
+                ) and all(isinstance(_node, SchedulerNode) for _node in __nodes)
+
+                # Step 1:
+                # Cross check each node between flatten_lazy_nodes_list and __nodes
+                for node1, node2 in itertools.product(flatten_lazy_nodes_list, __nodes):
+                    # If any node in __nodes is ancestors node for one of _lazy_nodes_list's
+                    if node2.get_names() & node1.ancestors:
+                        return False
```

Comment:
Since we do the `outer loop fusion` here, I think we also need the `will_fusion_create_cycle` check. We may consider of reusing the `will_fusion_create_cycle` here.